src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,500 @@
+                    if (ppvObject == IntPtr.Zero)
+                        return HResults.COR_E_INVALIDCAST;
+                }
+
+                AddRef();
+                return HResults.S_OK;
+            }
+
+            public IntPtr As(in Guid riid)
+            {
+                // Find target interface and return dispatcher or null if not found.
+                IntPtr typeMaybe = AsRuntimeDefined(in riid);
+                if (typeMaybe == IntPtr.Zero)
+                    typeMaybe = AsUserDefined(in riid);
+
+                return typeMaybe;
+            }
+
+            public unsafe void Destroy()
+            {
+                if (Target == IntPtr.Zero)
+                {
+                    return;
+                }
+
+                RuntimeImports.RhHandleFree(Target);
+                Target = IntPtr.Zero;
+            }
+
+            private unsafe IntPtr AsRuntimeDefined(in Guid riid)
+            {
+                if ((Flags & CreateComInterfaceFlags.CallerDefinedIUnknown) == CreateComInterfaceFlags.None)
+                {
+                    if (riid == IID_IUnknown)
+                    {
+                        return (IntPtr)(Dispatches + UserDefinedCount);
+                    }
+                }
+
+                return IntPtr.Zero;
+            }
+
+            private unsafe IntPtr AsUserDefined(in Guid riid)
+            {
+                for (int i = 0; i < UserDefinedCount; ++i)
+                {
+                    if (UserDefined[i].IID == riid)
+                    {
+                        return (IntPtr)(Dispatches + i);
+                    }
+                }
+
+                return IntPtr.Zero;
+            }
+        }
+
+        internal unsafe class ManagedObjectWrapperHolder
+        {
+            private ManagedObjectWrapper* _wrapper;
+
+            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper)
+            {
+                _wrapper = wrapper;
+            }
+
+            public unsafe IntPtr ComIp => _wrapper->As(in ComWrappers.IID_IUnknown);
+
+            ~ManagedObjectWrapperHolder()
+            {
+                // Release GC handle created when MOW was built.
+                _wrapper->Destroy();
+                Marshal.FreeCoTaskMem((IntPtr)_wrapper);
+            }
+        }
+
+        internal unsafe class NativeObjectWrapper
+        {
+            private IntPtr _externalComObject;
+
+            public NativeObjectWrapper(IntPtr externalComObject)
+            {
+                _externalComObject = externalComObject;
+            }
+
+            public uint AddRef()
+            {
+                return ((delegate* unmanaged<IntPtr, uint>)IUnknownVtbl[1])(_externalComObject);
+            }
+
+            public uint Release()
+            {
+                return ((delegate* unmanaged<IntPtr, uint>)IUnknownVtbl[2])(_externalComObject);
+            }
+
+            private IntPtr* IUnknownVtbl => (IntPtr*)_externalComObject;
```

Comment:
This does not return Vtbl. It would need an extra indirection to actually return Vtbl.

Also, this needs some basic tests.

It may be better to omit the RCW support from this PR (I think it was close to be ready for merge - before these changes), and do this in subsequent PR.